File: 0022-generate-parentheses/solution.java
class Solution {
    public void recur(int open,int close,int n,String curr,List<String> res){
        if(curr.length()==n*2){
            res.add(curr);
            return ;
        }
        if(open<n){
            recur(open+1,close,n,curr+'(',res);
        }
        if(close<open){
            recur(open,close+1,n,curr+')',res);
        }

    }
    public List<String> generateParenthesis(int n) {
        List<String> res= new ArrayList<>();
        recur(0,0,n,"",res);
        return res;
    }
}

File: 0022-generate-parentheses/solution.cpp
class Solution {
public:
    vector<string> res;
    void recur(int n,string cur){
        if(n==0){
            stack<char> stack;
            for(int i=0;i<cur.size();i++){
                if(cur[i]=='('){
                    stack.push('(');
                }
                else if(stack.empty()){
                    return ;
                }
                else{
                    stack.pop();
                }
            }
            if(stack.empty()){
                res.push_back(cur);
            }
        }else{
            recur(n-1,cur+'(');
            recur(n-1,cur+')');
        }
    }
    vector<string> generateParenthesis(int n) {
        recur(n*2,"");
        return res;
    }
};
